`TestListUnifiedResources_MixedAccess` flakiness

## Failure

#### Link(s) to logs

https://github.com/gravitational/teleport/actions/runs/14362900844/job/40268687471

#### Relevant snippet

```
auth_with_roles_test.go:6108: 
        	Error Trace:	/__w/teleport/teleport/lib/auth/auth_with_roles_test.go:6108
        	Error:      	"[Node:<Kind:"node" Version:"v2" Metadata:<Name:"tifa2" Namespace:"default" Labels:<key:"name" value:"mylabel" > Revision:"c7975bdc-725d-4e1e-89dd-39946d578d99" > Spec:<Hostname:"tifa2" Rotation:<Started:<seconds:-62135596800 > LastRotated:<seconds:-62135596800 > Schedule:<UpdateClients:<seconds:-62135596800 > UpdateServers:<seconds:-62135596800 > Standby:<seconds:-62135596800 > > > > >  Node:<Kind:"node" Version:"v2" Metadata:<Name:"tifa" Namespace:"default" Labels:<key:"name" value:"mylabel" > Revision:"515ee85a-7035-4afa-9eb4-58112bbe8d88" > Spec:<Hostname:"tifa" Rotation:<Started:<seconds:-62135596800 > LastRotated:<seconds:-62135596800 > Schedule:<UpdateClients:<seconds:-62135596800 > UpdateServers:<seconds:-62135596800 > Standby:<seconds:-62135596800 > > > > >  DatabaseServer:<Kind:"db_server" Version:"v3" Metadata:<Name:"tifa2" Namespace:"default" Revision:"e9bcb228-5874-4ad8-9c3a-6e6d5d2f7e9c" > Spec:<Version:"18.0.0-dev" Hostname:"_" HostID:"_" Rotation:<Started:<seconds:-62135596800 > LastRotated:<seconds:-62135596800 > Schedule:<UpdateClients:<seconds:-62135596800 > UpdateServers:<seconds:-62135596800 > Standby:<seconds:-62135596800 > > > Database:<Kind:"db" Version:"v3" Metadata:<Name:"name-5" Namespace:"default" > Spec:<Protocol:"_" URI:"_" AWS:<Redshift:<> RDS:<> ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<> RedshiftServerless:<> OpenSearch:<> DocumentDB:<> > GCP:<> Azure:<Redis:<> > TLS:<> AD:<> MySQL:<> MongoAtlas:<> Oracle:<> > Status:<AWS:<Redshift:<> RDS:<> ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<> RedshiftServerless:<> OpenSearch:<> DocumentDB:<> > MySQL:<> Azure:<Redis:<> > > > > Status:<> >  DatabaseServer:<Kind:"db_server" Version:"v3" Metadata:<Name:"cid" Namespace:"default" Revision:"120212b8-3363-4595-8ac5-09e826a84154" > Spec:<Version:"18.0.0-dev" Hostname:"_" HostID:"_" Rotation:<Started:<seconds:-62135596800 > LastRotated:<seconds:-62135596800 > Schedule:<UpdateClients:<seconds:-62135596800 > UpdateServers:<seconds:-62135596800 > Standby:<seconds:-62135596800 > > > Database:<Kind:"db" Version:"v3" Metadata:<Name:"name-4" Namespace:"default" > Spec:<Protocol:"_" URI:"_" AWS:<Redshift:<> RDS:<> ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<> RedshiftServerless:<> OpenSearch:<> DocumentDB:<> > GCP:<> Azure:<Redis:<> > TLS:<> AD:<> MySQL:<> MongoAtlas:<> Oracle:<> > Status:<AWS:<Redshift:<> RDS:<> ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<> RedshiftServerless:<> OpenSearch:<> DocumentDB:<> > MySQL:<> Azure:<Redis:<> > > > > Status:<> >  DatabaseServer:<Kind:"db_server" Version:"v3" Metadata:<Name:"baret" Namespace:"default" Revision:"5658cbd8-85c0-4742-8fa6-97b55c374d49" > Spec:<Version:"18.0.0-dev" Hostname:"_" HostID:"_" Rotation:<Started:<seconds:-62135596800 > LastRotated:<seconds:-62135596800 > Schedule:<UpdateClients:<seconds:-62135596800 > UpdateServers:<seconds:-62135596800 > Standby:<seconds:-62135596800 > > > Database:<Kind:"db" Version:"v3" Metadata:<Name:"name-3" Namespace:"default" > Spec:<Protocol:"_" URI:"_" AWS:<Redshift:<> RDS:<> ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<> RedshiftServerless:<> OpenSearch:<> DocumentDB:<> > GCP:<> Azure:<Redis:<> > TLS:<> AD:<> MySQL:<> MongoAtlas:<> Oracle:<> > Status:<AWS:<Redshift:<> RDS:<> ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<> RedshiftServerless:<> OpenSearch:<> DocumentDB:<> > MySQL:<> Azure:<Redis:<> > > > > Status:<> >  DatabaseServer:<Kind:"db_server" Version:"v3" Metadata:<Name:"aerith" Namespace:"default" Revision:"47ba6d46-ccb9-48d6-ab81-4b615b3f299e" > Spec:<Version:"18.0.0-dev" Hostname:"_" HostID:"_" Rotation:<Started:<seconds:-62135596800 > LastRotated:<seconds:-62135596800 > Schedule:<UpdateClients:<seconds:-62135596800 > UpdateServers:<seconds:-62135596800 > Standby:<seconds:-62135596800 > > > Database:<Kind:"db" Version:"v3" Metadata:<Name:"name-2" Namespace:"default" > Spec:<Protocol:"_" URI:"_" AWS:<Redshift:<> RDS:<> ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<> RedshiftServerless:<> OpenSearch:<> DocumentDB:<> > GCP:<> Azure:<Redis:<> > TLS:<> AD:<> MySQL:<> MongoAtlas:<> Oracle:<> > Status:<AWS:<Redshift:<> RDS:<> ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<> RedshiftServerless:<> OpenSearch:<> DocumentDB:<> > MySQL:<> Azure:<Redis:<> > > > > Status:<> >  DatabaseServer:<Kind:"db_server" Version:"v3" Metadata:<Name:"cloud" Namespace:"default" Revision:"2dbf7885-f733-4e5e-82de-30ba94f2b97b" > Spec:<Version:"18.0.0-dev" Hostname:"_" HostID:"_" Rotation:<Started:<seconds:-62135596800 > LastRotated:<seconds:-62135596800 > Schedule:<UpdateClients:<seconds:-62135596800 > UpdateServers:<seconds:-62135596800 > Standby:<seconds:-62135596800 > > > Database:<Kind:"db" Version:"v3" Metadata:<Name:"name-1" Namespace:"default" > Spec:<Protocol:"_" URI:"_" AWS:<Redshift:<> RDS:<> ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<> RedshiftServerless:<> OpenSearch:<> DocumentDB:<> > GCP:<> Azure:<Redis:<> > TLS:<> AD:<> MySQL:<> MongoAtlas:<> Oracle:<> > Status:<AWS:<Redshift:<> RDS:<> ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<> RedshiftServerless:<> OpenSearch:<> DocumentDB:<> > MySQL:<> Azure:<Redis:<> > > > > Status:<> >  DatabaseServer:<Kind:"db_server" Version:"v3" Metadata:<Name:"tifa" Namespace:"default" Revision:"4b38547a-a9f0-44d1-aadf-f04f54859778" > Spec:<Version:"18.0.0-dev" Hostname:"_" HostID:"_" Rotation:<Started:<seconds:-62135596800 > LastRotated:<seconds:-62135596800 > Schedule:<UpdateClients:<seconds:-62135596800 > UpdateServers:<seconds:-62135596800 > Standby:<seconds:-62135596800 > > > Database:<Kind:"db" Version:"v3" Metadata:<Name:"name-0" Namespace:"default" > Spec:<Protocol:"_" URI:"_" AWS:<Redshift:<> RDS:<> ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<> RedshiftServerless:<> OpenSearch:<> DocumentDB:<> > GCP:<> Azure:<Redis:<> > TLS:<> AD:<> MySQL:<> MongoAtlas:<> Oracle:<> > Status:<AWS:<Redshift:<> RDS:<> ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<> RedshiftServerless:<> OpenSearch:<> DocumentDB:<> > MySQL:<> Azure:<Redis:<> > > > > Status:<> >  Node:<Kind:"node" Version:"v2" Metadata:<Name:"cloud" Namespace:"default" Labels:<key:"name" value:"mylabel" > Revision:"5751980d-c298-48f3-8062-1dc1528623dc" > Spec:<Hostname:"cloud" Rotation:<Started:<seconds:-62135596800 > LastRotated:<seconds:-62135596800 > Schedule:<UpdateClients:<seconds:-62135596800 > UpdateServers:<seconds:-62135596800 > Standby:<seconds:-62135596800 > > > > >  Node:<Kind:"node" Version:"v2" Metadata:<Name:"cid" Namespace:"default" Labels:<key:"name" value:"mylabel" > Revision:"470e8983-60d2-4e5f-813c-c9dbdaacfd09" > Spec:<Hostname:"cid" Rotation:<Started:<seconds:-62135596800 > LastRotated:<seconds:-62135596800 > Schedule:<UpdateClients:<seconds:-62135596800 > UpdateServers:<seconds:-62135596800 > Standby:<seconds:-62135596800 > > > > >  Node:<Kind:"node" Version:"v2" Metadata:<Name:"baret" Namespace:"default" Labels:<key:"name" value:"mylabel" > Revision:"07556b12-0e0f-481b-bb3d-6f4e0cf77a03" > Spec:<Hostname:"baret" Rotation:<Started:<seconds:-62135596800 > LastRotated:<seconds:-62135596800 > Schedule:<UpdateClients:<seconds:-62135596800 > UpdateServers:<seconds:-62135596800 > Standby:<seconds:-62135596800 > > > > >  Node:<Kind:"node" Version:"v2" Metadata:<Name:"aerith" Namespace:"default" Labels:<key:"name" value:"mylabel" > Revision:"573f5016-ea8c-41e4-a045-8f362f7dfabb" > Spec:<Hostname:"aerith" Rotation:<Started:<seconds:-62135596800 > LastRotated:<seconds:-62135596800 > Schedule:<UpdateClients:<seconds:-62135596800 > UpdateServers:<seconds:-62135596800 > Standby:<seconds:-62135596800 > > > > > ]" should have 6 item(s), but has 12
        	Test:       	TestListUnifiedResources_MixedAccess
```

https://github.com/gravitational/teleport/actions/runs/14650307311/job/41114250912

I was able to reproduce this locally by changing the wait time in `flushCache` to 5ms.

@fspmarshall How reliable should `flushCache` be in tests? Does writing a new role through the cache guarantee that previous writes have completed, or are writes handled synchronously?

I also found a slightly different, likely related flake:

```
--- FAIL: TestListUnifiedResources_MixedAccess (1.10s)
    auth_with_roles_test.go:5869: 
        	Error Trace:	/Users/bjoerger/src/gravitational/teleport/lib/auth/auth_with_roles_test.go:5869
        	Error:      	Should be true
        	Test:       	TestListUnifiedResources_MixedAccess
        	Messages:   	Expected Access Denied, got <nil>
FAIL
FAIL	github.com/gravitational/teleport/lib/auth	6.350s
FAIL
```

--- a/lib/auth/auth_with_roles_test.go
+++ b/lib/auth/auth_with_roles_test.go
@@ -6152,7 +6152,7 @@ func TestListUnifiedResources_MixedAccess(t *testing.T) {
 		})
 		assert.True(t, trace.IsAccessDenied(err), "Expected Access Denied, got %v", err)
 		assert.Nil(t, resp)
-	}, 3*time.Second, 200*time.Millisecond)
+	}, 10*time.Second, 200*time.Millisecond)
 
 	// Validate that an error is returned when a subset of kinds are requested.
 	resp, err = clt.ListUnifiedResources(ctx, &proto.ListUnifiedResourcesRequest{

--- a/lib/auth/auth_with_roles_test.go
+++ b/lib/auth/auth_with_roles_test.go
@@ -6122,7 +6122,7 @@ func TestListUnifiedResources_MixedAccess(t *testing.T) {
 		assert.NoError(t, err)
 		assert.Empty(t, resp.NextKey)
 		assert.Len(t, resp.Resources, 6)
-	}, 3*time.Second, 200*time.Millisecond)
+	}, 10*time.Second, 200*time.Millisecond)
 
 	// only receive databases because nodes are denied with labels and desktops are denied with a verb rule
 	for _, resource := range resp.Resources {

--- a/lib/auth/auth_with_roles_test.go
+++ b/lib/auth/auth_with_roles_test.go
@@ -6114,14 +6114,14 @@ func TestListUnifiedResources_MixedAccess(t *testing.T) {
 	flushCache(t, srv.Auth())
 
 	var resp *proto.ListUnifiedResourcesResponse
-	inlineEventually(t, func() bool {
+	require.EventuallyWithT(t, func(t *assert.CollectT) {
 		resp, err = clt.ListUnifiedResources(ctx, &proto.ListUnifiedResourcesRequest{
 			Limit:  20,
 			SortBy: types.SortBy{IsDesc: true, Field: types.ResourceMetadataName},
 		})
-		require.NoError(t, err)
-		require.Empty(t, resp.NextKey)
-		return len(resp.Resources) == 6
+		assert.NoError(t, err)
+		assert.Empty(t, resp.NextKey)
+		assert.Len(t, resp.Resources, 6)
 	}, 3*time.Second, 200*time.Millisecond)
 
 	// only receive databases because nodes are denied with labels and desktops are denied with a verb rule
@@ -6145,12 +6145,13 @@ func TestListUnifiedResources_MixedAccess(t *testing.T) {
 	require.NoError(t, err)
 
 	// Validate that an error is returned when no kinds are requested.
-	inlineEventually(t, func() bool {
-		_, err = clt.ListUnifiedResources(ctx, &proto.ListUnifiedResourcesRequest{
+	require.EventuallyWithT(t, func(t *assert.CollectT) {
+		resp, err := clt.ListUnifiedResources(ctx, &proto.ListUnifiedResourcesRequest{
 			Limit:  20,
 			SortBy: types.SortBy{IsDesc: true, Field: types.ResourceMetadataName},
 		})
-		return trace.IsAccessDenied(err)
+		assert.True(t, trace.IsAccessDenied(err), "Expected Access Denied, got %v", err)
+		assert.Nil(t, resp)
 	}, 3*time.Second, 200*time.Millisecond)
 
 	// Validate that an error is returned when a subset of kinds are requested.

--- a/lib/auth/auth_with_roles_test.go
+++ b/lib/auth/auth_with_roles_test.go
@@ -6108,16 +6108,21 @@ func TestListUnifiedResources_MixedAccess(t *testing.T) {
 	clt, err := srv.NewClient(identity)
 	require.NoError(t, err)
 
-	// ensure updated roles have propagated to auth cache
+	// ensure updated roles have propagated to auth cache. Since ListUnifiedResources
+	// uses a separate, custom cache, this is not guaranteed to flush that cache, so
+	// we still use an Eventually loop below.
 	flushCache(t, srv.Auth())
 
-	resp, err := clt.ListUnifiedResources(ctx, &proto.ListUnifiedResourcesRequest{
-		Limit:  20,
-		SortBy: types.SortBy{IsDesc: true, Field: types.ResourceMetadataName},
-	})
-	require.NoError(t, err)
-	require.Len(t, resp.Resources, 6)
-	require.Empty(t, resp.NextKey)
+	var resp *proto.ListUnifiedResourcesResponse
+	inlineEventually(t, func() bool {
+		resp, err = clt.ListUnifiedResources(ctx, &proto.ListUnifiedResourcesRequest{
+			Limit:  20,
+			SortBy: types.SortBy{IsDesc: true, Field: types.ResourceMetadataName},
+		})
+		require.NoError(t, err)
+		require.Empty(t, resp.NextKey)
+		return len(resp.Resources) == 6
+	}, 3*time.Second, 200*time.Millisecond)
 
 	// only receive databases because nodes are denied with labels and desktops are denied with a verb rule
 	for _, resource := range resp.Resources {
@@ -6130,7 +6135,9 @@ func TestListUnifiedResources_MixedAccess(t *testing.T) {
 	_, err = srv.Auth().UpsertRole(ctx, role)
 	require.NoError(t, err)
 
-	// ensure updated roles have propagated to auth cache
+	// ensure updated roles have propagated to auth cache. Since ListUnifiedResources
+	// uses a separate, custom cache, this is not guaranteed to flush that cache, so
+	// we still use an Eventually loop below.
 	flushCache(t, srv.Auth())
 
 	// Get a new client to test with the new roles.
@@ -6138,13 +6145,13 @@ func TestListUnifiedResources_MixedAccess(t *testing.T) {
 	require.NoError(t, err)
 
 	// Validate that an error is returned when no kinds are requested.
-	resp, err = clt.ListUnifiedResources(ctx, &proto.ListUnifiedResourcesRequest{
-		Limit:  20,
-		SortBy: types.SortBy{IsDesc: true, Field: types.ResourceMetadataName},
-	})
-
-	require.True(t, trace.IsAccessDenied(err), "Expected Access Denied, got %v", err)
-	require.Nil(t, resp)
+	inlineEventually(t, func() bool {
+		_, err = clt.ListUnifiedResources(ctx, &proto.ListUnifiedResourcesRequest{
+			Limit:  20,
+			SortBy: types.SortBy{IsDesc: true, Field: types.ResourceMetadataName},
+		})
+		return trace.IsAccessDenied(err)
+	}, 3*time.Second, 200*time.Millisecond)
 
 	// Validate that an error is returned when a subset of kinds are requested.
 	resp, err = clt.ListUnifiedResources(ctx, &proto.ListUnifiedResourcesRequest{

